packages/lodestar/src/sync/backfill/backfill.ts

```diff
@@ -0,0 +1,189 @@
+
+export enum BackfillSyncStatus {
+  /** Required history blocks are syncing */
+  Syncing,
+
+  /** There are no suitable peers or we synced all required history blocks */
+  Idle,
+}
+
+type BackfillSyncState =
+  | {
+      status: BackfillSyncStatus.Syncing;
+      target: ChainTarget;
+    }
+  | {status: BackfillSyncStatus.Idle};
+
+export type RangeSyncModules = {
+  chain: IBeaconChain;
+  db: IBeaconDb;
+  network: INetwork;
+  config: IBeaconConfig;
+  logger: ILogger;
+};
+
+export type BackfillSyncOpts = BatchOpts;
+
+export class BackfillSync extends (EventEmitter as {new (): BackfillSyncEmitter}) {
+  private readonly chain: IBeaconChain;
+  private readonly network: INetwork;
+  private readonly db: IBeaconDb;
+  private readonly config: IBeaconConfig;
+  private readonly logger: ILogger;
+  private opts?: BackfillSyncOpts;
+
+  private syncChain?: SyncChain;
+
+  constructor(modules: RangeSyncModules, opts?: BackfillSyncOpts) {
+    super();
+    this.chain = modules.chain;
+    this.network = modules.network;
+    this.db = modules.db;
+    this.config = modules.config;
+    this.logger = modules.logger;
+    this.opts = opts;
+    void this.update();
+    this.network.events.addListener(NetworkEvent.peerConnected, this.addPeer);
+    this.network.events.addListener(NetworkEvent.peerDisconnected, this.removePeer);
+  }
+
+  /** Throw / return all AsyncGenerators inside every SyncChain instance */
+  close(): void {
+    this.network.events.removeListener(NetworkEvent.peerConnected, this.addPeer);
+    this.network.events.removeListener(NetworkEvent.peerDisconnected, this.removePeer);
+    this.syncChain?.remove();
+  }
+
+  state(): BackfillSyncState {
+    if (this.syncChain?.isSyncing && this.syncChain?.target) {
+      return {status: BackfillSyncStatus.Syncing, target: this.syncChain.target};
+    }
+    return {status: BackfillSyncStatus.Idle};
+  }
+
+  private async update(): Promise<void> {
+    const oldestStoredBlock = await this.db.blockArchive.firstValue();
+    if (!oldestStoredBlock) {
+      //no blocks in db, somethings wrong
+      return;
+    }
+    const oldestStoredBlockSlot = oldestStoredBlock.message.slot;
+    const oldestRequiredEpoch = Math.max(
+      GENESIS_EPOCH,
+      this.chain.clock.currentEpoch - getMinEpochForBlockRequests(this.config)
+    );
+    const oldestSlotRequired = computeStartSlotAtEpoch(this.config, oldestRequiredEpoch);
+    if (oldestSlotRequired >= oldestStoredBlockSlot) {
+      this.close();
+    } else {
+      this.syncChain?.startSyncing(oldestRequiredEpoch);
+    }
+  }
+
+  private async addPeer(peerId: PeerId, peerStatus: phase0.Status): Promise<void> {
+    //Could be caveat if you start from genesis, than stop and continue from WS state
+    // Currently we are relying on block archive prune to delete old blocks after which this is going to work
+    const oldestStoredBlock = await this.db.blockArchive.firstValue();
```

Comment:
We don't really need to check this for every new peer, but once at startup and that's it.